Mathlib/Algebra/Homology/HomotopyCategory/HomComplexShift.lean

```diff
@@ -0,0 +1,229 @@
+
+lemma rightShift_v (a n' : ℤ) (hn' : n' + a = n) (p q : ℤ) (hpq : p + n' = q)
+    (p' : ℤ) (hp' : p + n = p') :
+    (γ.rightShift a n' hn').v p q hpq = γ.v p p' hp' ≫
+      (L.shiftFunctorObjXIso a q p' (by rw [← hp', ← hpq, ← hn', add_assoc])).inv := by
+  subst hp'
+  dsimp only [rightShift]
+  simp only [mk_v]
+
+/-- The map `Cochain K (L⟦a⟧) n' → Cochain K L n` when `n' + a = n`. -/
+def rightUnshift {n' a : ℤ} (γ : Cochain K (L⟦a⟧) n') (n : ℤ) (hn : n' + a = n) :
+    Cochain K L n :=
+  Cochain.mk (fun p q hpq => γ.v p (p + n') rfl ≫
+    (L.shiftFunctorObjXIso a (p + n') q (by rw [← hpq, add_assoc, hn])).hom)
+
+lemma rightUnshift_v {n' a : ℤ} (γ : Cochain K (L⟦a⟧) n') (n : ℤ) (hn : n' + a = n)
+    (p q : ℤ) (hpq : p + n = q) (p' : ℤ) (hp' : p + n' = p') :
+    (γ.rightUnshift n hn).v p q hpq = γ.v p p' hp' ≫
+      (L.shiftFunctorObjXIso a p' q (by rw [← hpq, ← hn, ← add_assoc, hp'])).hom := by
+  subst hp'
+  dsimp only [rightUnshift]
+  simp only [mk_v]
+
+@[simp]
+lemma rightUnshift_rightShift (a n' : ℤ) (hn' : n' + a = n) :
+    (γ.rightShift a n' hn').rightUnshift n hn' = γ := by
+  ext p q hpq
+  simp only [rightUnshift_v _ n hn' p q hpq (p + n') rfl,
+    γ.rightShift_v _ _ hn' p (p + n') rfl q hpq,
+    shiftFunctorObjXIso, assoc, Iso.inv_hom_id, comp_id]
+
+@[simp]
+lemma rightShift_rightUnshift {a n' : ℤ} (γ : Cochain K (L⟦a⟧) n') (n : ℤ) (hn' : n' + a = n) :
+    (γ.rightUnshift n hn').rightShift a n' hn' = γ := by
+  ext p q hpq
+  simp only [(γ.rightUnshift n hn').rightShift_v a n' hn' p q hpq (p + n) rfl,
+    γ.rightUnshift_v n hn' p (p + n) rfl q hpq,
+    shiftFunctorObjXIso, assoc, Iso.hom_inv_id, comp_id]
+
+variable (K L n)
+
+@[simp]
+lemma rightShift_zero (a n' : ℤ) (hn' : n' + a = n) :
+    (0 : Cochain K L n).rightShift a n' hn' = 0 := by
+  ext p q hpq
+  dsimp
+  rw [rightShift_v _ a n' hn' p q hpq _ rfl, zero_v, zero_comp]
+
+@[simp]
+lemma rightUnshift_zero (a n' : ℤ) (hn' : n' + a = n) :
+    (0 : Cochain K (L⟦a⟧) n').rightUnshift n hn' = 0 := by
+  ext p q hpq
+  dsimp
+  rw [rightUnshift_v _ n _ _ _ _ (q - a) (by linarith), zero_v, zero_comp]
+
+variable {K L n}
+
+@[simp]
+lemma rightShift_neg (a n' : ℤ) (hn' : n' + a = n) :
+    (-γ).rightShift a n' hn' = -γ.rightShift a n' hn' := by
+  ext p q hpq
+  dsimp
+  simp only [rightShift_v _ a n' hn' p q hpq _ rfl, neg_v, neg_comp]
+
+@[simp]
+lemma rightShift_add (a n' : ℤ) (hn' : n' + a = n) :
+    (γ₁ + γ₂).rightShift a n' hn' = γ₁.rightShift a n' hn' + γ₂.rightShift a n' hn' := by
+  ext p q hpq
+  dsimp
+  simp only [rightShift_v _ a n' hn' p q hpq _ rfl, add_v, add_comp]
+
+variable (K L)
+
+/-- The additive equivalence `Cochain K L n ≃+ Cochain K L⟦a⟧ n'` when `n' + a = n`. -/
+@[simps]
+def rightShiftAddEquiv (n a n' : ℤ) (hn' : n' + a = n) :
+    Cochain K L n ≃+ Cochain K (L⟦a⟧) n' where
+  toFun γ := γ.rightShift a n' hn'
+  invFun γ := γ.rightUnshift n hn'
+  left_inv γ := by simp
+  right_inv γ := by simp
+  map_add' γ γ' := by simp
```

Comment:
I think you should move this up a bit higher. And then you can deduce all the lemmas about preserving `0` and `-` etc, by just restating the goal in terms of this AddEquiv, and applying some generic `map_zero` or `map_neg` lemma.

Also... should this even be upgraded to a linear equiv?

Reply:
Thanks for the suggestion! I have deduced the lemmas that can be deduced from the additive equivalence, and I have also made a linear equivalence version.